src/Compilers/VisualBasic/Portable/Symbols/Source/SourceModuleSymbol.vb

```diff
@@ -887,19 +887,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                                                symbol As Symbol) As Boolean
             Debug.Assert(Me.DeclaringCompilation.EventQueue IsNot Nothing)
 
+            Dim change As Boolean = False
             SyncLock _diagnosticLock
-                Dim change = (variable And mask) = comparand
+                change = (variable And mask) = comparand
                 If change Then
-                    Me.DeclaringCompilation.SymbolDeclaredEvent(symbol)
-
                     If Not ThreadSafeFlagOperations.Set(variable, mask) Then
                         ' If this gets hit, then someone wrote to variable without going through this
                         ' routine, which is a bug.
                         Throw ExceptionUtilities.Unreachable
                     End If
                 End If
-                Return change
             End SyncLock
+
+            If change Then
+                ' Invoke the symbol declared event outside the lock.
+                ' Task scheduler might decide to inline the event processing, which can call back into analyzers,
```

Comment:
@gafter I left a comment [here](https://github.com/dotnet/roslyn/pull/6171#discussion_r42444161) on the methods being invoked during AsyncQueue.Enqueue - I presume we didn't expect TaskCompletionSource.SetResult to eventually invoke FinishContinuations?